src/histogram.rs

```diff
@@ -0,0 +1,401 @@
+            }
+            Bound::PosInf => false,
+        }
+    }
+}
+
+/// The Histogram type.
+///
+/// TODO more docs to come 
+#[derive(Debug)]
+pub struct Histogram<T>
+where
+    T: Copy,
+{
+    count: usize,
+    sum: Option<T>,
+    bins: Vec<(Bound<T>, usize)>,
+}
+
+#[derive(Debug, Copy, Clone)]
+/// Construction errors
+///
+/// `Histogram` is a little finicky when you construct it. We signal errors out
+/// to the user with this enumeration.
+pub enum Error {
+    /// The bounds given to Histogram are empty. We need bounds.
+    BoundsEmpty,
+    /// The bounds given to Histogram are not sorted. They must be.
+    BoundsNotSorted,
+}
+
+fn is_sorted<T>(bounds: &[T]) -> bool
+where
+    T: cmp::PartialOrd + fmt::Debug,
+{
+    let mut prev = None;
+    for i in bounds {
+        if prev.is_none() {
+            prev = Some(i);
+            continue;
+        }
+        let p: &T = prev.unwrap();
+        match i.partial_cmp(p) {
+            Some(cmp::Ordering::Less) => {
+                return false;
+            }
+            _ => {
+                prev = Some(i);
+            }
+        }
+    }
+    true
+}
+
+impl<T> Histogram<T>
+where
+    T: Copy + cmp::PartialOrd + fmt::Debug,
+{
+    /// TODO documentation
+    pub fn new(bounds: Vec<T>) -> Result<Histogram<T>, Error> {
+        if bounds.is_empty() {
+            return Err(Error::BoundsEmpty);
+        }
+        if !is_sorted(&bounds) {
+            return Err(Error::BoundsNotSorted);
+        }
+
+        let mut bins: Vec<(Bound<T>, usize)> = bounds
+            .into_iter()
+            .map(|x| (Bound::Finite(x), usize::min_value()))
+            .collect();
+        let cap: (Bound<T>, usize) = (Bound::PosInf, 0);
+        bins.push(cap);
+
+        Ok(Histogram {
+            count: 0,
+            sum: None, // TODO make summations, tests for same 
+            bins: bins,
+        })
+    }
+
+    /// TODO documentation
+    pub fn insert(&mut self, value: T) -> () {
+        let mut idx = 0;
+        let val_bound = Bound::Finite(value);
+        for &(ref bound, _) in &self.bins {
+            match bound.partial_cmp(&val_bound) {
+                Some(cmp::Ordering::Greater) | Some(cmp::Ordering::Equal) => {
+                    break;
+                }
+                Some(cmp::Ordering::Less) | None => idx += 1,
+            }
+        }
+        self.bins[idx].1 += 1;
+        self.count += 1;
+    }
+
+    /// Returns the total number of items stored in the histogram 
+    /// 
+    /// TODO documentation
+    pub fn count(&self) -> usize {
+        self.count
+    }
+
+    /// Total number of items below supplied upper_bound
+    ///
+    /// TODO documentation
+    pub fn total_below(&self, upper: Bound<T>) -> usize {
+        let mut count = 0;
+        for &(ref bound, cnt) in &self.bins {
+            if bound > &upper {
+                break;
+            } else {
+                count += cnt;
+            }
+        }
+        count
+    }
```

Comment:
Doesn't this give the wrong result if I construct a histogram with a set of bounds and then do total_below with a bound not in the set given to the constructor? Should it require upper to lie on one of the pre-defined boundaries? (Same question for the others below)

Another thought that struck me is that we can make this faster if the bins are cumulative, then you just need to locate the right one and return that value, and the others become simpler too.